src/aspire/classification/rir_class2d.py

```diff
@@ -0,0 +1,577 @@
+        # Also most third party NN complain about complex...
+        #   so we'll pretend we have 2*n_features of real values.
+        # Don't worry about the copy because NearestNeighbors wants
+        #   C-contiguous anyway... (it would copy internally otherwise)
+        X = np.column_stack((coeff_b.real, coeff_b.imag))
+        # We'll also want to consider the neighbors under reflection.
+        #   These coefficients should be provided by coeff_b_r
+        X_r = np.column_stack((coeff_b_r.real, coeff_b_r.imag))
+
+        # We can compare both non-reflected and reflected representations as one large set by
+        #   taking care later that we store refl=True where indices>=n_img
+        X_both = np.concatenate((X, X_r))
+
+        nbrs = NearestNeighbors(n_neighbors=self.n_nbor, algorithm="auto").fit(X_both)
+        distances, indices = nbrs.kneighbors(X)
+
+        # any refl?
+        logger.info(
+            f"Count reflected: {np.sum(indices>=n_img)}"
+            f" ({np.sum(indices>=n_img) / len(indices)}%)"
+        )
+
+        if diagnostics:
+            # Lets peek at the distribution of distances
+            plt.hist(
+                distances[:, 1:].flatten(), bins="auto"
+            )  # zero index is self, distance 0
+            plt.show()
+
+        # I was considering to change the result of these functions
+        # but for now return classes as range 0 to 2*n_img.
+        # #########################
+        # # There are two sets of vectors each n_img long.
+        # #   The second set represents reflected.
+        # #   When a reflected coef vector is a nearest neighbor,
+        # #   we notate the original image index (indices modulus),
+        # #   and notate we'll need the reflection (refl).
+        # classes = indices % n_img
+        # refl = np.array(indices // n_img, dtype=bool)
+        # corr = distances
+        # return classes, refl, corr
+
+        return indices
+
+    def eigen_images(self):
+        """
+        Return the eigen images of the FSPCA basis, evaluated to image space.
+
+        This may be used to implot visualizations of the eigenvectors.
+
+        Ordering corresponds to FSPCA eigvals.
+        """
+
+        eigvecs = self.pca_basis.eigvecs
+        if isinstance(eigvecs, BlkDiagMatrix):
+            eigvecs = eigvecs.dense()
+
+        return self.fb_basis.evaluate(eigvecs.T)
+
+    def output(self, classes, classes_refl, rot, coefs=None, include_refl=True):
+        """
+        Return class averages.
+
+        :param classes: class indices (refering to src). (n_img, n_nbor)
+        :param classes_refl: Bool representing whether to reflect image in `classes`
+        :param rot: Array represting totation angle (Radians) of image in `classes`
+        :coefs: Optional Fourier bessel coefs (avoids recomputing).
+        :return: Stack of Synthetic Class Average images as Image instance.
+        """
+
+        if not include_refl:
```

Comment:
When is this needed? Is it just for debugging?

Reply:
Mainly for debugging. It has also been my personal experience that the (alignment) results are better in low res without introducing the reflected set, but tbh, this method works much better at higher resolutions afaict, so I'll just remove it.